src/bin/pgcopydb/catalog.c

```diff
@@ -7170,6 +7222,113 @@ catalog_iter_s_table_in_copy_init(SourceTableIterator *iter)
 }
 
 
+/*
+ * catalog_iter_s_index_not_processed iterates over the list of indexes which
+ * are not processed in our catalogs. This might be due to user interruption.
+ */
+bool
+catalog_iter_s_index_not_processed(DatabaseCatalog *catalog,
+								   void *context,
+								   SourceIndexIterFun *callback)
+{
+	SourceIndexIterator *iter =
+		(SourceIndexIterator *) calloc(1, sizeof(SourceIndexIterator));
+
+	iter->catalog = catalog;
+
+	if (!catalog_iter_s_index_not_processed_init(iter))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	for (;;)
+	{
+		if (!catalog_iter_s_index_next(iter))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		SourceIndex *index = iter->index;
+
+		if (index == NULL)
+		{
+			if (!catalog_iter_s_index_finish(iter))
+			{
+				/* errors have already been logged */
+				return false;
+			}
+
+			break;
+		}
+
+		/* now call the provided callback */
+		if (!(*callback)(context, index))
+		{
+			log_error("Failed to iterate over list of indexs, "
+					  "see above for details");
+			return false;
+		}
+	}
+
+
+	return true;
+}
+
+
+/*
+ * catalog_iter_s_index_not_processed_init initializes an Interator over our
+ * catalog of SourceIndex entries which are yet to be processed or interrupted
+ * during the processing.
+ */
+bool
+catalog_iter_s_index_not_processed_init(SourceIndexIterator *iter)
+{
+	sqlite3 *db = iter->catalog->db;
+
+	if (db == NULL)
+	{
+		log_error("BUG: Failed to initialize s_index iterator: db is NULL");
+		return false;
+	}
+
+	iter->index = (SourceIndex *) calloc(1, sizeof(SourceIndex));
+
+	if (iter->index == NULL)
+	{
+		log_error(ALLOCATION_FAILED_ERROR);
+		return false;
+	}
+
+	char *sql =
+		"  select i.oid, i.qname, i.nspname, i.relname, i.restore_list_name, "
+		"         i.tableoid, t.qname, t.nspname, t.relname, "
+		"         isprimary, isunique, columns, i.sql, "
+		"         c.oid as constraintoid, conname, "
+		"         condeferrable, condeferred, c.sql as condef"
+		"    from s_index i "
+		"         join s_table t on t.oid = i.tableoid "
+		"         left join s_constraint c on c.indexoid = i.oid"
+		"   where i.oid not in ("
+		"	      select indexoid from summary where"
+		"				done_time_epoch is not null	and indexoid is not null)";
```

Comment:
It's best to use WHERE NOT EXISTS rather than NOT IN, because of SQL semantics with NULL in the list and its impact of performance characteristics of query execution. Please rewrite the WHERE clause?